src/main/java/net/dries007/tfc/common/container/KnappingContainer.java

```diff
@@ -0,0 +1,199 @@
+
+    public KnappingContainer(ContainerType<?> containerType, IRecipeType<? extends KnappingRecipe> recipeType, int windowId, PlayerInventory playerInv, int amountToConsume, boolean consumeAfterComplete, boolean usesDisabledTex, boolean needsKnife, SoundEvent sound)
+    {
+        super(containerType, windowId, playerInv, getItemForKnapping(playerInv), 20);
+        this.itemIndex += 1;
+        this.amountToConsume = amountToConsume;
+        this.usesDisabledTex = usesDisabledTex;
+        this.consumeAfterComplete = consumeAfterComplete;
+        this.needsKnife = needsKnife;
+        this.recipeType = recipeType;
+        this.sound = sound;
+
+        matrix = new SimpleCraftMatrix();
+        hasBeenModified = false;
+        requiresReset = false;
+        hasConsumedIngredient = false;
+        stackCopy = this.stack.copy();
+    }
+
+    private static ItemStack getItemForKnapping(PlayerInventory inv)
+    {
+        final ItemStack main = inv.player.getMainHandItem();
+        return main.isEmpty() ? inv.player.getOffhandItem() : main;
+    }
+
+    @Override
+    public void onButtonPress(int buttonID, @Nullable CompoundNBT extraNBT)
+    {
+        matrix.set(buttonID, false);
+
+        if (!hasBeenModified)
+        {
+            if (!player.isCreative() && !consumeAfterComplete)
+            {
+                ItemStack consumedStack = Helpers.consumeItem(this.stack, amountToConsume);
+                if (isOffhand)
+                {
+                    player.setItemInHand(Hand.OFF_HAND, consumedStack);
+                }
+                else
+                {
+                    player.setItemInHand(Hand.MAIN_HAND, consumedStack);
+                }
+            }
+            hasBeenModified = true;
+        }
+
+        // check the pattern
+        Slot slot = slots.get(0);
+        if (slot != null && player.level instanceof ServerWorld)
+        {
+            KnappingRecipe recipe = getMatchingRecipe((ServerWorld) player.level);
+            if (recipe != null)
+            {
+                slot.set(recipe.assemble(this));
+            }
+            else
+            {
+                slot.set(ItemStack.EMPTY);
+            }
+        }
+    }
+
+    @Override
+    protected void addContainerSlots()
+    {
+        addSlot(new RunnableSlot(new ItemStackHandler(1), 0, 128, 46, this::resetMatrix));
+    }
+
+    @Override
+    public void removed(PlayerEntity player)
+    {
+        Slot slot = slots.get(0);
+        ItemStack stack = slot.getItem();
+        if (!stack.isEmpty())
+        {
+            if (!player.level.isClientSide)
+            {
+                ItemHandlerHelper.giveItemToPlayer(player, stack);
+                consumeIngredientStackAfterComplete();
+            }
+        }
+        super.removed(player);
+    }
+
+    /**
+     * Used in client to check a slot state in the matrix
+     * JEI won't cause issues anymore see https://github.com/TerraFirmaCraft/TerraFirmaCraft/issues/718
```

Comment:
I'd remove this issue reference since the code in question is basically being rewritten.